Add support for using `PF_Twitter` without a linking or logging into a Parse account

zadr opened the pull request:
There are two things going on here, in two separate commits:
1. On 16121e21f9f0b34555f4ef24183949d832e3983d, expose/document how `-[PFTwitter init]` works in that it is okay to call, and after setting `consumerKey` and `consumerSecret`, can be used to make Twitter API calls.
2. In 6e5e9ca6381f15428bfc579ce201143ab631b9b7, add support for deauthorization, by calling `oauth2/invalidate_token` on Twitter's API, and then to clear out local state on success.

6e5e9ca6381f15428bfc579ce201143ab631b9b7 contains tests for the newly added `deauthorizeInBackground` call, however, I wasn't able to get the tests to run locally, so, I can't promise that the tests do the right thing (after running `bundle install` locally, updating the deployment target, enabling ARC in the project and a few other steps that have slipped my mind, the rabbit hole seemed too deep to keep following).

Went to sign the CLA, and it told me `The 'Email' value has been registered on this form before. If you would like to update an existing contributor account, please email cla@fb.com with your details.`, so, I think thats all taken care of.

Commit:
Document the behavior of `-[PF_Twitter init]`

It is useful to be able to create a standalone instance of `PFTwitter` to make authenticated Twitter API calls without having to link a Twitter account to a Parse account or require Twitter login.

Commit:
Add support for deauthorizing from Twitter to invalidate session tokens

nlutsenko commented on ParseTwitterUtils/PF_Twitter.h:
```diff
@@ -22,6 +22,15 @@ NS_ASSUME_NONNULL_BEGIN
 @interface PF_Twitter : NSObject
 
 /**
+ An instance of `PF_Twitter` configured to access device Twitter accounts with Accounts.framework,
```
Nit: `Initializes an instance of ...`

nlutsenko commented on ParseTwitterUtils/PF_Twitter.m:
```diff
@@ -92,6 +92,30 @@ - (BFTask *)authorizeInBackground {
+
+    return [[self _performDeauthAsync] pftw_continueAsyncWithBlock:^id(BFTask *task) {
+        BFTaskCompletionSource *source = [BFTaskCompletionSource taskCompletionSource];
+        dispatch_async(dispatch_get_main_queue(), ^{
```
Not sure whether we need to actually dispatch to main thread here, since the model of tasks is to move to another thread on the next task, but not on this one.
This basically means - that you don't need to `dispatch_async` here, and can simply `setLoginResultValues:`

zadr replied:
👍 will remove -- and confession: this was mostly copied from `authorizeInBackground`, which also has a `dispatch_async` to the main queue (and why theres a `//TODO` with your name in it, oops, heh). worth removing that `dispatch_async` at the same time, because of the same logic?

nlutsenko commented on ParseTwitterUtils/PF_Twitter.m:
```diff
@@ -92,6 +92,30 @@ - (BFTask *)authorizeInBackground {
+            } else if (task.error) {
+                [source trySetError:task.error];
+            } else if (task.result) {
+                [self setLoginResultValues:@{}];
```
Since we are reading from all the fields and reading is safe from `nil` - you can simply pass `nil` here?
Doesn't change the behavior, but avoids creating another dictionary.

zadr replied:
I went with passing in an empty dictionary literal since it's nice to not have an optional when dealing with Swift, and, well, it's a cheap operation (iirc, it's like one extra `mov` instruction compared to passing in `nil`).

that said, i don't feel that too strongly about this, and can revert if you would prefer ¯_(ツ)_/¯

zadr replied:
thinking about it, i'll just make it `nil`, since, it matches what the rest of the project does, and thats the more important thing here anyway. sorry for those few stubborn minutes 😅

nlutsenko commented on Tests/Unit/TwitterTests.m:
```diff
@@ -622,4 +622,67 @@ - (void)testAuthorizeWithoutLocalAccountAndNetworkSuccess {
+        XCTAssertNil(error);
+        XCTAssertNotNil(task.result);
+
+		XCTAssertNil(twitter.userId);
```
Nit: Please fix indentation on lines 677-680

nlutsenko replied:
Owe, also this looks like this logic is redundant, since you need to check for `authToken`/`authTokenSecret` as well.

zadr replied:
👍 good point! if we don't have an auth token/secret, we shouldn't have any info about the current twitter account anyway. will change it around.

nlutsenko commented on Tests/Unit/TwitterTests.m:
```diff
@@ -622,4 +622,67 @@ - (void)testAuthorizeWithoutLocalAccountAndNetworkSuccess {
+        NSError *error = task.error;
+        XCTAssertNotNil(error);
+        XCTAssertEqualObjects(error.domain, PFParseErrorDomain);
+        XCTAssertEqual(error.code, 2);
```
This needs to have an error code of `1`. Causes a test to fail :)

zadr replied:
oops! a previous iteration of this commit added a new error code of `2`, and I missed this in cleaning it up 😅

nlutsenko commented:
Hey @zadr, this is amazing and almost flawless.
Most of the requested changes are small nits, but the most important part is actually to fix tests, since we don't want to merge in something that fails to build :)

I am not sure on why OCMock doesn't like that expectation and doesn't want to mock it, so this would need a little bit of debugging probably. If you can't make it work though - you can simply create an issue for it to implement it later and I'll be happy to jump on it at any given point.
The other approach that we can take is I could patch it locally and see what I can dig up. Let me know which route would be more helpful for you, so we can unblock it and merge it in!

Overall, once again, this looks great and is an amazing idea to implement this!

Commit:
code review feedback, and fix tests

zadr commented:
@nlutsenko thanks for 👀ing this so quickly and thoroughly! added a new commit that addresses all your nits, and, should make all the tests pass.

Using Xcode 7.3, what I had to do to get the tests run (and see what was failing) by: explicitly set the deployment target of the xcodeproj to iphoneos, and then running the following command:
```
xcodebuild test -workspace ParseTwitterUtils.xcworkspace -sdk iphonesimulator -scheme ParseTwitterUtils-iOS -configuration Debug GCC_INSTRUMENT_PROGRAM_FLOW_ARCS=YES GCC_GENERATE_TEST_COVERAGE_FILES=YES CLANG_ENABLE_OBJC_ARC=YES
```

Which differs from the command Travis uses to run with the tests on Xcode 7.2.1:
```
xcodebuild test -workspace ParseTwitterUtils.xcworkspace -sdk iphonesimulator -scheme ParseTwitterUtils-iOS -configuration Debug -destination "platform=iOS Simulator,name=iPhone 4s" -destination "platform=iOS Simulator,name=iPhone 6s" GCC_INSTRUMENT_PROGRAM_FLOW_ARCS=YES GCC_GENERATE_TEST_COVERAGE_FILES=YES
```
in that it doesn't specify the `-destination`, and adds `CLANG_ENABLE_OBJC_ARC=YES` at the end.

Earlier, on another machine, I wasn't able to get the same steps running with Xcode 7.3.1, so, I guess, Apple changed things around again?

nlutsenko commented:
`CLANG_ENABLE_OBJC_ARC=YES` shouldn't be required... Maybe you didn't checkout the submodules? The pure difference comes from the fact that we want to test both 32 and 64 bit versions of code (that's why 2 destinations). It is not required for development though, since you can simply `Test` the framework target, which will run all the tests ;)

It passes as a PR on Travis - that's the important part. You can ignore `codecov` there, since those guys changed a bunch of things and it still is semi-stable.

nlutsenko commented on Tests/Unit/TwitterTests.m:
```diff
@@ -622,4 +622,62 @@ - (void)testAuthorizeWithoutLocalAccountAndNetworkSuccess {
+        NSError *error = task.error;
+        XCTAssertNil(error);
+        XCTAssertNotNil(task.result);
+
```
The checks for `userId` and `screenName` were still useful, sorry if I was not explicit enough about this portion. I feel like it's fine to leave it as is, but if you want to update it till next morning (will merge it then) - feel free to update this part.

nlutsenko commented:
Looks great. Will merge in tomorrow morning.

Commit:
Re-add nil XCTAsserts for userId and screenName during deauth test

zadr commented:
Re-added the asserts.

And, hm. There aren't any submodules in the projects, and, I had to run `pod install` to get `OCMock`. I'll assume I'm just missing something obvious 'cos of the time, and not worry about it, though!

To run tests in Xcode, (opening up  `ParseTwitterUtils.xcworkspace` and hitting ⌘U), I had to make the same changes to the project to enable ARC and to set the Base SDK (not deployment target, as I said earlier, whoops) to Latest iOS SDK before they'd run.

Thanks again for reviewing this so quickly (and so late^wearly, this time :D)

nlutsenko commented:
We actually do have a submodule there in `Vendor` folder for `xctoolchain` which contains the common `xcconfig` files for all of the configurations (including enabling ARC).
Try running `git submodule update --init --recursive` in your local clone :grin:

zadr commented:
(The obvious thing i'm missing is that `git submodule update` doesn't do anything until `git submodule update --init` is run first)

nlutsenko commented:
Perfect, merging now. Just sent a PR as well for updating all dependencies and Travis CI to Xcode 7.3, so we are going to get better on that front as well.
